source/common/formatter/substitution_formatter.h

```diff
@@ -253,8 +255,192 @@ template <class FormatterContext> class FormatterBaseImpl : public FormatterBase
+   * formatters. Then join the raw string and output of substitution formatters in order
+   * to get the final output.
+   *
+   * @param struct_format the proto struct format configuration.
+   */
+  FormatterEelements fromStruct(const ProtobufWkt::Struct& struct_format) const;
+
+private:
+  void formatValueToFormatterEelements(const ProtobufWkt::Struct& value) const;
+  void formatValueToFormatterEelements(const ProtobufWkt::Value& value) const;
+
+  const bool keep_value_type_{};
+
+  mutable std::string sanitize_buffer_; // Buffer for sanitizing strings.
+  mutable std::string raw_json_string_; // Buffer for joining raw strings.
+  mutable FormatterEelements output_;   // Final output.
+};
+
+template <class FormatterContext>
+class JsonFormatterImplBase : public FormatterBase<FormatterContext> {
+public:
+  using CommandParsers = std::vector<CommandParserBasePtr<FormatterContext>>;
+  using Formatters = std::vector<FormatterProviderBasePtr<FormatterContext>>;
+
+  JsonFormatterImplBase(const ProtobufWkt::Struct& struct_format, bool keep_value_type,
+                        bool omit_empty_values, const CommandParsers& commands = {})
+      : omit_empty_values_(omit_empty_values), keep_value_type_(keep_value_type),
+        empty_value_(omit_empty_values_ ? std::string()
+                                        : std::string(DefaultUnspecifiedValueStringView)) {
+    const auto elements = JsonFormatBuilder(keep_value_type).fromStruct(struct_format);
+    for (const auto& element : elements) {
+      if (absl::holds_alternative<JsonFormatBuilder::TmplString>(element)) {
+        elements_.emplace_back(SubstitutionFormatParser::parse<FormatterContext>(
+            absl::get<JsonFormatBuilder::TmplString>(element).value_, commands));
+      } else {
+        elements_.emplace_back(absl::get<JsonFormatBuilder::JsonString>(element).value_);
+      }
+    }
+  }
+
+  std::string formatWithContext(const FormatterContext& context,
+                                const StreamInfo::StreamInfo& info) const {
+    std::string log_line;
+    log_line.reserve(512);
+
+    std::string sanitize_buffer;
+
+    for (const auto& element : elements_) {
+      // 1. Handle the raw string element.
+      if (absl::holds_alternative<std::string>(element)) {
+        log_line.append(absl::get<std::string>(element));
+        continue;
+      }
+      ASSERT(absl::holds_alternative<Formatters>(element));
+
+      const auto& formatters = absl::get<Formatters>(element);
+      ASSERT(!formatters.empty());
+
+      // 2. Handle the formatter element with multiple providers or case
+      //    that value type needs not to be kept.
+      if (formatters.size() > 1 || !keep_value_type_) {
+        // Multiple providers forces string output.
+        log_line.push_back('"');
+        for (const auto& provider : formatters) {
+          const auto bit = provider->formatWithContext(context, info);
```

Comment:
why the variable name `bit` -- is this. 0 or 1? If so, why sanitize?

Reply:
I actually don't know. This naming is kept for consistence with exist code. I don't know why.
I will change it if you think it's fine.